docs/Brew-Livecheck.md

```diff
@@ -0,0 +1,115 @@
+
+* **Use a symbol for a formula URL (i.e. `:homepage`, `:stable`, and `:head`) when appropriate**, to avoid duplicating formula URLs in the livecheckable.
+
+* **It's generally preferable to check for versions in the same location as the stable archive, when possible**. This preference is stronger for first-party sources (websites, repositories, etc.) and becomes weaker for third-party sources (e.g. mirrors, another software package manager, etc.).
+
+### Regex guidelines
+
+The regex in a `livecheck` block is used within a strategy to restrict matching to only relevant text (or strings) and to establish what part of the matched text is the version (using a capture group).
+
+Creating a good regex is a balance between being too strict (breaking easily) and too loose (matching more than it should).
+
+* For technical reasons, **the `regex` call in the `livecheck` block should always use parentheses** (e.g. `regex(/example/)`).
+
+* **Regex literals should follow the established Homebrew style**, where the `/.../` syntax is the default and the `%r{...}` syntax is used when a forward slash (`/`) is present.
+
+* **Regexes should adequately represent their intention.** This requires an understanding of basic regex syntax, so we avoid issues like using `.` (match any character) instead of `\.` when we only want to match a period. We also try to be careful about our use of generic catch-alls like  `.*` or `.+`, as it's often better to use something non-greedy and contextually appropriate. For example, if we wanted to match a variety of characters while trying to stay within the bounds of an HTML attribute, we could use something like `[^"' >]*?`.
+
+* **Try not to be too specific in some parts of the regex.** For example, if a file name uses a hyphen (`-`) between the software name and version (e.g. `example-1.2.3.tar.gz`), we may want to use something like `example[._-]v?(\d+(?:\.\d+)+)\.t` instead. This would allow the regex to continue matching if the upstream file name format changes to `example.1.2.3.tar.gz` or `example_1.2.3.tar.gz`.
+
+* **Regexes should be case insensitive unless case sensitivity is explicitly required for matching to work properly**. Case insensitivity is enabled by adding the `i` flag at the end of the regex literal (e.g. `/.../i` or `%r{...}i`). This helps to improve reliability and reduce maintenance, as a case-insensitive regex doesn't need to be manually updated if there are any upstream changes in letter case.
+
+* **Regexes should only use a capturing group around the part of the matched text that corresponds to the version**. For example, in `/href=.*?example-v?(\d+(?:\.\d+)+)(?:-src)?\.t/i`, we're only using a capturing group around the version part (matching a version like `1.2`, `1.2.3`, etc.) and we're using non-capturing groups elsewhere (e.g. `(?:-src)?`). This allows livecheck to rely on the first capture group being the version string.
```

Comment:
Why would you need a second capture group?

Reply:
Alternatively: if the second capture group is unused anyway: why does this matter as long as the first one is the one matching the version?

Reply:
I've massaged this language a bit to assert that capturing groups should only be used around version text, as there seemed to be some confusion around the "first capture group" language.

> Why would you need a second capture group?

In the overwhelming majority of cases, a regex only contains one capture group: the version text. We now have a small number of regexes that contain more than one capture group, where the regex is used within a `strategy` block to combine the captured text into one version (e.g., a cask version like `1.2.3, abcdefgh`).

> Alternatively: if the second capture group is unused anyway: why does this matter as long as the first one is the one matching the version?

It's one of those situations where it's fine until it's not. It's easier and more reliable to say, "Only use a capture group around the version text" instead of "you can use a capture group as long as it comes after all the capture group(s) for version text". The latter case would allow people to be lazy about this rule in most cases but it would come back to bite us.

For a practical example, let's say an upstream repository historically tags versions like `1.2.3`. The standard regex for Git tags like this (`/^v?(\d+(?:\.\d+)+)$/i`) only has a capture group around the version, so it works fine. Over time, upstream switches to a `release-1.2.3` tag format but they're not always reliable about it, so we end up with `1.2.3`, `release-1.2.4`, `1.2.5`, and `release-1.2.6`.

Since we have to match both of these tag formats, we would need to add an optional prefix at the start of the existing regex. If we're following the strict guideline, we would use `/^(?:release[._-])?v?(\d+(?:\.\d+)+)$/i`, which uses a non-capturing group around the optional `release-` part. This works as expected.

If we aren't strict about this rule, someone may naively use a regex like `/^(release[._-])?v?(\d+(?:\.\d+)+)$/i`, which would treat `release-` as the version. This `release-` version is technically newer than `1.2.5` (due to how `Version` comparison works) and this check would pass CI since livecheck technically found a version. If a maintainer doesn't manually test the `livecheck` block outside of CI (I always do) or check the CI log to catch this, it would be merged and break the check.

---

That said, I believe we have to be strict about this anyway to be able to enforce it using a RuboCop. From an implementation standpoint, I imagine this RuboCop will basically be "ensure there's only one capture group in the regex when there isn't a `strategy` block".